Yatima/FromLean.lean

```diff
@@ -1,186 +1,308 @@
+    return .max univACid univBCid
+  | .imax a b _  => do
+    let univA ← toYatimaUniv lvls a
+    let univACid ← univToCid univA
+    addToEnv $ .univ_cache univACid univA
+    let univB ← toYatimaUniv lvls b
+    let univBCid ← univToCid univB
+    addToEnv $ .univ_cache univBCid univB
+    return .imax univACid univBCid
+  | .param nam _ => match lvls.indexOf nam with
+    | some n => return .param nam n
+    | none   => throw s!"'{nam}' not found in '{lvls}'"
+  | .mvar .. => throw "Unfilled level metavariable"
+
+  partial def toYatimaRecursorRule
+    (ctorCid : ConstCid) (rules : Lean.RecursorRule) :
+      EnvM RecursorRule := do
+    let rhs ← toYatimaExpr [] rules.rhs
+    let rhsCid ← exprToCid rhs
+    addToEnv $ .expr_cache rhsCid rhs
+    return ⟨ctorCid, rules.nfields, rhsCid⟩
+
+  partial def toYatimaExpr (levelParams : List Lean.Name) :
+      Lean.Expr → EnvM Expr
+    | .bvar idx _ => return .var "" idx
+    | .sort lvl _ => do
+      let univ ← toYatimaUniv levelParams lvl
+      addToEnv $ .univ_cache (← univToCid univ) univ
+      return .sort (← toYatimaUniv levelParams lvl)
+    | .const nam lvls _ => do
+      match (← read).find?' nam with
+      | some leanConst =>
+        let const ← toYatimaConst leanConst
+        let constId ← constToCid const
+        addToEnv $ .const_cache constId const
+        let univs ← lvls.mapM $ toYatimaUniv levelParams
+        for univ in univs do
+          addToEnv $ .univ_cache (← univToCid univ) univ
+        return .const nam constId univs
+      | none => throw s!"Unknown constant '{nam}'"
+    | .app fnc arg _ => do
+      let fnc ← toYatimaExpr levelParams fnc
+      addToEnv $ .expr_cache (← exprToCid fnc) fnc
+      let arg ← toYatimaExpr levelParams arg
+      addToEnv $ .expr_cache (← exprToCid arg) arg
+      return .app fnc arg
+    | .lam nam bnd bod _ => do
+      let bndInfo := bnd.binderInfo
+      let bnd ← toYatimaExpr levelParams bnd
+      addToEnv $ .expr_cache (← exprToCid bnd) bnd
+      let bod ← toYatimaExpr levelParams bod
+      addToEnv $ .expr_cache (← exprToCid bod) bod
+      return .lam nam bndInfo bnd bod
+    | .forallE nam dom img _ => do
+      let bndInfo := dom.binderInfo
+      let dom ← toYatimaExpr levelParams dom
+      addToEnv $ .expr_cache (← exprToCid dom) dom
+      let img ← toYatimaExpr levelParams img
+      addToEnv $ .expr_cache (← exprToCid img) img
+      return .pi nam bndInfo dom img
+    | .letE nam typ exp bod _ => do
+      let typ ← toYatimaExpr levelParams typ
+      addToEnv $ .expr_cache (← exprToCid typ) typ
+      let exp ← toYatimaExpr levelParams exp
+      addToEnv $ .expr_cache (← exprToCid exp) exp
+      let bod ← toYatimaExpr levelParams bod
+      addToEnv $ .expr_cache (← exprToCid bod) bod
+      return .letE nam typ exp bod
+    | .lit lit _ => return Yatima.Expr.lit lit
+    | .mdata _ e _ => toYatimaExpr levelParams e
+    | .proj .. => sorry
+    | .fvar .. => sorry
```

Comment:
There shouldn't be any free variables (`fvar`) or metavariables (`mvar`) when we compile from Lean, so I think this should error. I don't really know what `proj` does and haven't been able to find a good explanation of that in the docs. Would be great to investigate that further